inst/include/cpp11/R.hpp

```diff
@@ -33,6 +33,13 @@ constexpr R_xlen_t operator"" _xl(unsigned long long int value) { return value;
 
 }  // namespace literals
 
+namespace traits {
+template <typename T>
+struct get_underlying_type {
+  using type = T;
+};
+}  // namespace traits
```

Comment:
The `traits` namespace may be overkill, I'm not sure. 

I have seen this idea used elsewhere though. One of those places is Rcpp:
https://github.com/RcppCore/Rcpp/blob/master/inst/include/Rcpp/traits/storage_type.h